crates/biome_js_analyze/src/analyzers/nursery/no_duplicate_test_hooks.rs

```diff
@@ -0,0 +1,221 @@
+    /// ### Valid
+    ///
+    /// ```js
+    /// describe('foo', () => {
+    ///   beforeEach(() => {
+    ///     // some setup
+    ///   });
+    ///   test('foo_test', () => {
+    ///     // some test
+    ///   });
+    /// });
+    /// ```
+    ///
+    pub(crate) NoDuplicateTestHooks {
+        version: "next",
+        name: "noDuplicateTestHooks",
+        recommended: true,
+    }
+}
+
+#[derive(Debug, Default)]
+struct HooksContext {
+    before_all: usize,
+    before_each: usize,
+    after_all: usize,
+    after_each: usize,
+}
+
+#[derive(Default)]
+struct DuplicateHooksVisitor {
+    stack: Vec<HooksContext>,
+}
+
+impl Visitor for DuplicateHooksVisitor {
+    type Language = JsLanguage;
+
+    fn visit(
+        &mut self,
+        event: &WalkEvent<SyntaxNode<Self::Language>>,
+        mut ctx: VisitorContext<Self::Language>,
+    ) {
+        match event {
+            WalkEvent::Enter(node) => {
+                // When the visitor enters a function node, push a new entry on the stack
+                if let Some(node) = JsCallExpression::cast_ref(node) {
+                    if let Ok(callee) = node.callee() {
+                        if callee.contains_a_test_pattern().is_ok() {
+                            if let Some(function_name) = get_function_name(&callee) {
+                                if function_name.text_trimmed() == "describe" {
+                                    self.stack.push(HooksContext::default());
+                                }
+                            }
+                        }
+                    }
+                }
+
+                if let Some(node) = JsCallExpression::cast_ref(node) {
+                    if let Ok(AnyJsExpression::JsIdentifierExpression(identifier)) = node.callee() {
+                        identifier
+                            .name()
+                            .and_then(|name| name.value_token())
+                            .map_or((), |name| {
+                                if let Some(hooks_context) = self.stack.last_mut() {
+                                    match name.text_trimmed() {
+                                        "beforeEach" | "beforeAll" | "afterEach" | "afterAll" => {
+                                            let counter = match name.text_trimmed() {
+                                                "beforeEach" => &mut hooks_context.before_each,
+                                                "beforeAll" => &mut hooks_context.before_all,
+                                                "afterEach" => &mut hooks_context.after_each,
+                                                "afterAll" => &mut hooks_context.after_all,
+                                                _ => {
+                                                    unreachable!()
+                                                } // Should never happen
+                                            };
+                                            *counter += 1;
```

Comment:
It would be more readable to add a function to `HooksContext` called `add(&mut self, hook_name: &str) -> usize` that returns the new index. 

Also, let's add:
- a `debug_assert!` that checks that `name.text_trimmed()` is one of the strings you expect (inside this new function)

Reply:
Why `debug_assert` inside the function when we already matched for hook names?

Reply:
We can simply replace it with return 0 then?